The Comprehensive Guide to Elixir's List Comprehension (#186)
mhanberg/blog: _posts/2022-04-15-the-comprehensive-guide-to-elixirs-for-comprehension.md

  },
  %{
    name: "Mitch",
    status: :former,
    hobbies: [%{name: "Woodworking", type: :making}, %{name: "Homebrewing", type: :making}]
  },
  %{
    name: "Greg",
    status: :active,
    hobbies: [
      %{name: "Dungeons & Dragons", type: :gaming},
      %{name: "Woodworking", type: :making}
    ]
  }
]

base_map = %{
  "Mitch" => %{
    name: "Reading",
    type: :learning
  },
  "Greg" => %{
    name: "Traveling",
    type: :expensive
  }
}

for employee <- employees,
    employee.status == :active,
    hobby <- employee.hobbies,
    hobby.type == :gaming,
    into: base_map do
  {employee.name, hobby}
end

# %{
#   "Eric" => %{name: "Text Adventures", type: :gaming},
#   "Greg" => %{name: "Dungeons & Dragons", type: :gaming},
#   "Mitch" => %{name: "Reading", type: :learning}
# }
```

Here we can observe three things.

- The comprehension evaluates to a map.
- The `"Mitch"` key and its value were preserved in the final output.
- The `"Greg"` key's value in the `base_map` was overwritten by the value yielded during the comprehension with the same key. If our comprehension were to have returned multiple key/value pairs with identical keys, the last one would have won.

> This option is very useful for transforming maps, since iterating over a map with an `Enum` function turns it into a list of 2-tuples, you always need to pipe the return value into `Enum.into/2` or `Map.new/1`.

Reviewer: ```suggestion
This option is very useful for transforming maps. Since iterating over a map with an `Enum` function turns it into a list of 2-tuples, you always need to pipe the return value into `Enum.into/2` or `Map.new/1`.
```

First, the `:reduce` option takes a value that is to be used as the first value of the accumulator.

Second, the comprehension in this mode includes a slightly different syntax. Here the inside of the block includes the "arg(s) and right arrow" syntax that you see in anonymous functions and case expressions. This is the syntax that allows the comprehensions to yield the accumulator to the block on every iteration.

The additional syntax is the same as the other places you have probably seen it, you can pattern match and pass additional clauses!

```elixir
directions = [
  left: 2,
  up: 1,
  down: 5,
  right: 6
]

# You can't move below or to the left of 0.
starting_position = {0, 0}

for {dir, movement} <- directions, reduce: starting_position  do
  {x, y} when dir == :left and x - movement > 0 ->
    {x - movement, y}

  {x, y} when dir == :down and y - movement > 0 ->
    {x, y - movement}

  {x, y} when dir == :up ->
    {x, y + movement}

  {x, y} when dir == :right ->
    {x + movement, y}

  position ->
    IO.puts("Not possible to move #{dir} by #{movement} when you care located at #{inspect(position)}")

    position
end

# {6, 1}
```

> Above we can observe that we've written 5 different clauses, pattern match on the shape of the data, as well as add guard clauses that capture the data in the generator as well as the accumulator.

Reviewer: ```suggestion
Above we can observe that we've written 5 different clauses, pattern matched on the shape of the data, as well as added guard clauses that capture the data in the generator and the accumulator.
```

The beauty of using a comprehension as a reducer is the ability to use multiple generators and act on them as if they are one level of iteration.

```elixir
friends = [
  %{name: "Derek", hobbies: ["Movies", "Hot Sauce"]},
  %{name: "Joe", hobbies: ["Yu-Gi-Oh!", "Tattoos"]},
  %{name: "Andres", hobbies: ["Photoshop", "Oreos", "Cereal"]},
]

for %{hobbies: hobbies, name: name} <- friends, hobby <- hobbies, reduce: [] do
  tagged_hobbies ->
    [{name, hobby} | tagged_hobbies]
end

# [
#   {"Andres", "Cereal"},
#   {"Andres", "Oreos"},
#   {"Andres", "Photoshop"},
#   {"Joe", "Tattoos"},
#   {"Joe", "Yu-Gi-Oh!"},
#   {"Derek", "Hot Sauce"},
#   {"Derek", "Movies"}
# ]
```

To write this without a comprehension it would look something like:

```elixir
Enum.reduce(friends, [], fn %{name: name, hobbies: hobbies}, tagged_hobbies ->
  new_hobbies = Enum.map(hobbies, fn hobby -> {name, hobby} end)

  new_hobbies ++ tagged_hobbies
end)
```

## Conclusion

If you've made it this far, congrats! The comprehension packs a lot of features into a tiny programming construct and demonstrating all of them is a lot of work!

> The comprehension is one of my favorite features of the Elixir programming language and it was a pleasure to write about every feature in as much depth as I could.

Reviewer: ```suggestion
The comprehension is one of my favorite features of the Elixir programming language, and it was a pleasure to write about every feature in as much depth as I could.
```
